win32ss/user/ntuser/ghost.c

```diff
@@ -128,6 +219,109 @@ HWND FASTCALL UserHungWindowFromGhostWindow(HWND hwndGhost)
     return IntHungWindowFromGhostWindow(pGhostWnd);
 }
 
+static BOOL IntHaveToQuitGhosting(void)
+{
+    return !!NtUserFindWindowEx(NULL, NULL, &GhostClass, NULL, 0);
+}
+
+static ULONG NTAPI
+GhostThreadProc(PVOID Param)
+{
+    HWND hwndTarget = (HWND)Param;
+    MSG msg;
+
+    if (!IntCreateGhostWindow(hwndTarget))
+    {
+        DPRINT1("Unable to create ghost\n");
+        goto Quit;
+    }
+
+    // thread message loop
+    while (NtUserGetMessage(&msg, NULL, 0, 0))
+    {
+        switch (msg.message)
+        {
+            case PM_CREATE_GHOST:
+            {
+                hwndTarget = (HWND)msg.wParam;
+                if (!IntCreateGhostWindow(hwndTarget))
+                {
+                    DPRINT1("Unable to create ghost\n");
+                }
+                break;
+            }
+        }
+        NtUserTranslateMessage(&msg, 0);
+        NtUserDispatchMessage(&msg);
+
+        if (IntHaveToQuitGhosting())
+        {
+            DPRINT1("Have to quit ghost thread\n");
+            break;
+        }
+    }
+
+Quit:
+    RtlFreeHeap(GlobalUserHeap, 0, gGhostInfo);
+
+    NtSetEvent(gGhostInfo->GhostQuitEvent, NULL);
+
+    gGhostInfo = NULL;
+
+    RtlExitUserThread(0);
+
+    return 0;
+}
+
+BOOL IntAddGhost(HWND hwndTarget)
+{
+    NTSTATUS Status;
+    CLIENT_ID ClientId;
+    HANDLE GhostThreadHandle;
+
+    if (gGhostInfo)
+    {
+        return NtUserPostThreadMessage(gGhostInfo->ThreadId, PM_CREATE_GHOST,
+                                       (WPARAM)hwndTarget, 0);
+    }
+
+    gGhostInfo = RtlAllocateHeap(GlobalUserHeap, HEAP_ZERO_MEMORY, sizeof(GHOST_INFO));
+    if (!gGhostInfo)
+        return FALSE;
+
+    gGhostInfo->hwndTarget = hwndTarget;
+
+    // create events
+    Status = NtCreateEvent(&gGhostInfo->GhostStartupEvent, EVENT_ALL_ACCESS,
+                           NULL, SynchronizationEvent, FALSE);
+    Status = NtCreateEvent(&gGhostInfo->GhostQuitEvent, EVENT_ALL_ACCESS,
+                           NULL, SynchronizationEvent, FALSE);
+
+    // create thread
+    Status = RtlCreateUserThread(NtCurrentProcess(),
```

Comment:
Pretty sure a user thread can't execute code in win32k. Look at PsCreateSystemThread instead.